aten/src/ATen/mps/IndexKernels.h

```diff
@@ -0,0 +1,134 @@
+                              constant const void    * indexStrides  [[buffer(2)]],
+                              constant const uint3   * offsets       [[buffer(3)]],
+                              constant const void    * inputData     [[buffer(4)]],
+                              device void          * outputData    [[buffer(5)]],
+                              uint thread_index [[thread_position_in_grid]]);
+template
+[[host_name("index_select_short")]]
+kernel void index_select<short>(constant const IndexAB & indexAB       [[buffer(0)]],
+                                constant const void    * indexSizes    [[buffer(1)]],
+                                constant const void    * indexStrides  [[buffer(2)]],
+                                constant const uint3   * offsets       [[buffer(3)]],
+                                constant const void    * inputData     [[buffer(4)]],
+                                device void          * outputData    [[buffer(5)]],
+                                uint thread_index [[thread_position_in_grid]]);
+template
+[[host_name("index_select_char")]]
+kernel void index_select<char>(constant const IndexAB & indexAB       [[buffer(0)]],
+                               constant const void    * indexSizes    [[buffer(1)]],
+                               constant const void    * indexStrides  [[buffer(2)]],
+                               constant const uint3   * offsets       [[buffer(3)]],
+                               constant const void    * inputData     [[buffer(4)]],
+                               device void          * outputData    [[buffer(5)]],
+                               uint thread_index [[thread_position_in_grid]]);
+template
+[[host_name("index_select_uchar")]]
+kernel void index_select<uchar>(constant const IndexAB & indexAB       [[buffer(0)]],
+                                constant const void    * indexSizes    [[buffer(1)]],
+                                constant const void    * indexStrides  [[buffer(2)]],
+                                constant const uint3   * offsets       [[buffer(3)]],
+                                constant const void    * inputData     [[buffer(4)]],
+                                device void          * outputData    [[buffer(5)]],
+                                uint thread_index [[thread_position_in_grid]]);
+
+template
+[[host_name("index_select_bool")]]
+kernel void index_select<bool>(constant const IndexAB & indexAB       [[buffer(0)]],
+                               constant const void    * indexSizes    [[buffer(1)]],
+                               constant const void    * indexStrides  [[buffer(2)]],
+                               constant const uint3   * offsets       [[buffer(3)]],
+                               constant const void    * inputData     [[buffer(4)]],
+                               device void          * outputData    [[buffer(5)]],
+                               uint thread_index [[thread_position_in_grid]]);
```

Comment:
One doesn't need template per datatype, but rather template per elementsize (unless there is a memcpy like function on metal, than templates aren't needed at all)
```suggestion
template
[[host_name("index_select_32bit")]]
kernel void index_select<int>(constant const IndexAB & indexAB       [[buffer(0)]],
                                constant const void    * indexSizes    [[buffer(1)]],
                                constant const void    * indexStrides  [[buffer(2)]],
                                constant const uint3   * offsets       [[buffer(3)]],
                                constant const void    * inputData     [[buffer(4)]],
                                device void          * outputData    [[buffer(5)]],
                                uint thread_index [[thread_position_in_grid]]);
template
[[host_name("index_select_16bit")]]
kernel void index_select<short>(constant const IndexAB & indexAB       [[buffer(0)]],
                               constant const void   * indexSizes    [[buffer(1)]],
                               constant const void   * indexStrides  [[buffer(2)]],
                               constant const uint3  * offsets       [[buffer(3)]],
                               constant const void   * inputData     [[buffer(4)]],
                               device void         * outputData    [[buffer(5)]],
                               uint thread_index [[thread_position_in_grid]]);
template
[[host_name("index_select_64bit")]]
kernel void index_select<long>(constant const IndexAB & indexAB       [[buffer(0)]],
                               constant const void    * indexSizes    [[buffer(1)]],
                               constant const void    * indexStrides  [[buffer(2)]],
                               constant const uint3   * offsets       [[buffer(3)]],
                               constant const void    * inputData     [[buffer(4)]],
                               device void          * outputData    [[buffer(5)]],
                               uint thread_index [[thread_position_in_grid]]);
template
[[host_name("index_select_8bit")]]
kernel void index_select<char>(constant const IndexAB & indexAB       [[buffer(0)]],
                               constant const void    * indexSizes    [[buffer(1)]],
                               constant const void    * indexStrides  [[buffer(2)]],
                               constant const uint3   * offsets       [[buffer(3)]],
                               constant const void    * inputData     [[buffer(4)]],
                               device void          * outputData    [[buffer(5)]],
                               uint thread_index [[thread_position_in_grid]]);
```

Reply:
The 64/32/8 bit templates make sense for indexing, I'll update that. Regarding templates, we'd still need them - there is no memcpy in metal

Reply:
@malfet , can we follow this up in next PR, with `index_put`. We didn't want to club these two as it will become a huge PR.

Reply:
Sure, can you please file a followup issue?